keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+pre-provisioning internally. Kubernetes wouldn't know how to match
+pre-provisioned resources against claims because it has no understanding about the
+parameters.
+
+Allocation of a resource happens either immediately when a ResourceClaim gets
+created (“immediate allocation”) or when when a Pod is getting scheduled which
+needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec. Pods reference resource
+claims in a new PodSpec.Resources list. Each resource in that list
+then can be made available to one or more containers in that Pod.
+
+Immediate allocation is useful when allocating a resource is expensive (for
+example, programming an FPGA) and the resource therefore is meant to be used by
+multiple different Pods, either in parallel or one after the other. The
+downside is that Pod resource requirements cannot be considered when choosing
+where to allocate. If a Pod then cannot run on the node (or nodes, for
+network-attached resources) where the resource is available because resources
+like RAM or CPU are exhausted there, then the Pod also cannot run elsewhere.
+Delayed allocation solves this by integrating allocation with Pod scheduling.
+
+Allocation must be complete before a Pod is allowed to be scheduled onto a
+node. This avoids scenarios where a Pod is permanently assigned to a node which
+can't fit the pod because of the pod's other resource requirements.
+
+When a PodTemplateSpec in an app controller spec references a ResourceClaim by
+name, all Pods created by that controller also use that name and thus share the
+resources allocated for that ResourceClaim. This might be supported by the
+resource driver and can be useful, depending on the characteristics of the
+resource.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.Resources list
+can also be a ResourceClaimTemplate. When a Pod gets created, such a template
+will be used to create a normal ResourceClaim with the Pod as owner, and then
+the normal allocation of the resource takes place.
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. For delayed allocation, a node is selected tentatively
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+The resources allocated for a ResourceClaim can be shared by multiple
+containers in a pod. Depending on the capabilities defined in the
```

Comment:
the capabilities are opaque to the user though, right?

what happens if the pod author attempts to use the resource in a way that is inconsistent with those capabilities?

Reply:
> the capabilities are opaque to the user though, right?

These capabilities need to be document for a resource driver, together with the parameters that it supports. The wording is a bit too abstract here: there's really only capability that currently matters, the number of concurrent users.

> what happens if the pod author attempts to use the resource in a way that is inconsistent with those capabilities?

If they attempt to use the same ResourceClaim in too many pods, only some of them are allowed to run.

Reply:
In the meantime we simplified that to "at most one pod per resource" and "unlimited number of pods", but the effect is the same.